### PR TITLE
Document build commands and add guardrails to make build to prevent common problems

### PR DESCRIPTION
**Pull request by ScottSuarez**
Adding documentation of build commands for more transparency about how the build process works. 

I've avoid documenting every branching rule as there are some build rules that could be misinterpreted and result in issues or confusion. So I'm only documented the commands I expect uses to run..

Added prebuild validation to ensure that `OUTPUT_PATH` is a directory and `goimports` is installed.

closes https://github.com/hashicorp/terraform-provider-google/issues/13147

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: GNUmakefile [hunk 1]
```diff
@@ -91,6 +125,10 @@ upgrade-dcl:
 		sed ${SED_I} "s!.*declarative-resource-client-library.*!$$MOD_LINE!" go.mod.erb; echo "$$SUM_LINE" >> go.sum
 
 
+validate_environment:
```
Comment (melinath) [inline GNUmakefile:128]: This seems similar to `mmv1/tools/doctor` - would it be possible to include the OUTPUT_PATH check there?

**Review by melinath — CHANGES_REQUESTED** [review]
I'm a little concerned about documenting this here - not all contributors will think to check the makefile for docs. Could you add this information to the main docs site? There may also be some overlap with the existing doctor script.

File: GNUmakefile (hunk 1 above)
Comment (ScottSuarez) [reply GNUmakefile:128]: doctor doesn't run as a part of initial build. It wouldn't make sense to include output path there as doctor is run to diagnose dependency issues not check build parameter settings.

**Comment by ScottSuarez** [comment]
Let me look at the documentation and see if/where it can fit in there.. I do worry about overloading information within that section. I want to keep it relevant as the user needs to understand it.

File: GNUmakefile (hunk 1 above)
Comment (melinath) [reply GNUmakefile:128]: 🤦 yes, that makes sense. However, the doctor might be possible to run instead of the check for goimports, since it already includes that (and a few other things)

**Comment by melinath** [comment]
would it be possible to make the OUTPUT_PATH default to the standard value so that users (generally) don't have to provide it? And/or allow specifying these values as env vars instead of inline, so that users can set them in `.profile` or similar?

**Comment by ScottSuarez** [comment]
> would it be possible to make the OUTPUT_PATH default to the standard value based on the version being generated so that users (generally) don't have to provide it? 

I've thought about this but received a bit of push back on make things too autonomous from team members. If I made it too intelligent team members expressed fear of it doing things they wouldn't be aware of.

>And/or allow specifying these values as env vars instead of inline, so that users can set them in `.profile` or similar?

They already use environment variables! You should be able to utilize them that way.

File: GNUmakefile (hunk 1 above)
Comment (ScottSuarez) [reply GNUmakefile:128]: I'll take a look at running doctor.. I like the current solution since it's minimal, very fast and only outputs in case of failed detection. I'll take a look.

Comment (ScottSuarez) [reply GNUmakefile:96]: updated

File: docs/layouts/shortcodes/build_table.html [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  </style>
+  <table class="tg">
+  <thead>
+    <tr>
+      <th class="tg-c3ow" colspan="2"><span style="font-weight:bold">magic-modules - build commands</span></th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td class="tg-0pky"><span style="font-weight:bold">make</span><br>runs make build</td>
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:19]: `make`, `make build`, `make terraform` are all aliases; we should choose one and only document that one to make things easier / clearer for users who are just getting started. My initial thought would be `make terraform` because it's unambiguous - but it's also from a time when we were supporting multiple downstreams that weren't necessarily terraform-related. We could instead do `make provider` or similar. What do you think?

File: docs/layouts/shortcodes/build_table.html [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      <td class="tg-0pky"><span style="font-weight:bold">make build</span><br>  Builds the terraform provider with both mmv1 and tpgtools. Copies all generated files as a merge with the provider in OUTPUT_PATH.</td>
+    </tr>
+    <tr>
+      <td class="tg-0pky"><span style="font-weight:bold">make upgrade-dcl</span><br> A command to upgrade the version of the dcl. Used within the terraform provider and tpgtools.</td>
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:23]: upgrade-dcl should be documented in a new section about tpgtools when we add one rather than in the getting started guide. I'd prefer to keep that in a separate PR if that's all right.

File: docs/layouts/shortcodes/build_table.html [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      <td class="tg-0pky"><span style="font-weight:bold">make validate_environment</span><br> A prestep to make build to ensure that the build environment is suitable for a build.</td>
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:24]: Since validate_environment is already part of `make terraform` we probably don't need to document it separately on this page. What would you think about recommending this in the "set up your environment" page instead of the doctor script?

I remember talking previously about the output path / version validation being part of this make command vs not - I don't remember the conclusion though!

Separate but related topic: What is the use case for having validate_environment as a standalone command vs. doing the check directly in the generation command & having the doctor script? The main benefit I see immediately is that `make validate_environment` would let us change the implementation easily later. But it is a bit long - what about `make doctor`? (Analogous to [`brew doctor`](https://docs.brew.sh/Troubleshooting) for homebrew.)

File: docs/layouts/shortcodes/build_table.html [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    </tr>
+    <tr>
+      <td class="tg-c3ow" colspan="2"><span style="font-weight:bold">magic-modules - build parameters</span></td>
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:27]: This part of the table seems to duplicate some (but not all) of the content from first section of this page. It would be great to combine the information so that it is only stated once.

**Review by melinath — CHANGES_REQUESTED** [review]

File: docs/content/docs/getting-started/generate-providers.md [hunk 6]
```diff
@@ -1,9 +1,11 @@
 ---
-title: Generate the providers
+title: Generating the provider(s)
```
[the remark this answers is not in the document]
Comment (ScottSuarez) [reply docs/content/docs/getting-started/generate-providers.md:2]: acc, makes sense to me !

File: docs/layouts/shortcodes/build_table.html (hunk 4 above)
Comment (ScottSuarez) [reply docs/layouts/shortcodes/build_table.html:24]: ack this makes sense.  I'll omit in favor of just running doctor (already supported). Sure updating validate_environment makes sense to me!

File: GNUmakefile [hunk 7]
```diff
@@ -91,6 +93,12 @@ upgrade-dcl:
...
+# only print doctor script to console if there was a depenecy failure detected.
+	@./scripts/doctor 2>&1 > /dev/null || ./scripts/doctor
+	@[ -d "${OUTPUT_PATH}" ] || (printf " \e[1;31mdirectory '${OUTPUT_PATH}' does not exist - ENV variable \033[0mOUTPUT_PATH\e[1;31m should be set to a provider directory... \033[0m \n" && exit 1);
+	@[ -d "${VERSION}" ] || (printf " \e[1;31mversion '${VERSION}' does not exist - ENV variable \033[0mVERSION\e[1;31m should be set to ga or beta \033[0m \n" && exit 1);
```
[the remark this answers is not in the document]
Comment (ScottSuarez) [reply GNUmakefile:100]: strange ! let me see..

File: docs/layouts/shortcodes/build_table.html (hunk 3 above)
Comment (ScottSuarez) [reply docs/layouts/shortcodes/build_table.html:23]: sounds good!

File: docs/layouts/shortcodes/build_table.html (hunk 5 above)
Comment (ScottSuarez) [reply docs/layouts/shortcodes/build_table.html:22]: removed the restated content.

File: docs/content/docs/getting-started/generate-providers.md [hunk 8]
```diff
@@ -1,9 +1,10 @@
 ---
-title: Generate the providers
+title: Generate the provider
 weight: 20
 ---
 
-# Generate the providers
+## Generating the provider(s)
```
Comment (melinath) [inline docs/content/docs/getting-started/generate-providers.md:6]: This should match whatever is in the title for the page.

```suggestion
# Generate the providers
```

File: docs/content/docs/getting-started/generate-providers.md [hunk 9]
```diff
@@ -1,9 +1,10 @@
 ---
-title: Generate the providers
+title: Generate the provider
```
Comment (melinath) [inline docs/content/docs/getting-started/generate-providers.md:2]: Ah - it had an `s` before - I don't have a preference, do you?

```suggestion
title: Generate the providers
```

File: GNUmakefile [hunk 10]
```diff
@@ -54,7 +55,8 @@ endif
 ifeq ($(FORCE_DCL),)
   FORCE_DCL=latest
 endif
-terraform build:
+terraform build provider:
+	@make validate_environment;
```
Comment (melinath) [inline GNUmakefile:59]: what is the benefit of having this be a separate call vs having the same lines directly in here?

File: docs/content/docs/getting-started/generate-providers.md [hunk 11]
```diff
@@ -32,6 +33,10 @@ to reduce the generation time.
 make terraform VERSION=ga OUTPUT_PATH="$GOPATH/src/github.com/hashicorp/terraform-provider-google" PRODUCT=foo
```
Comment (melinath) [inline docs/content/docs/getting-started/generate-providers.md:33]: if we're standardizing on `make provider` then we should replace all references to `make terraform` in our docs with `make provider` instead. We should update the container-based process as well so that people don't get confused about the discrepancy if they look in the Dockerfile.

File: docs/content/docs/getting-started/generate-providers.md [hunk 12]
```diff
@@ -32,6 +33,10 @@ to reduce the generation time.
...
 ```
 
+## Build command guide
```
Comment (melinath) [inline docs/content/docs/getting-started/generate-providers.md:36]:
```suggestion
## `make provider` reference
```

File: docs/content/docs/getting-started/generate-providers.md [hunk 13]
```diff
@@ -32,6 +33,10 @@ to reduce the generation time.
...
+
+{{< build_table >}}
```
Comment (melinath) [inline docs/content/docs/getting-started/generate-providers.md:38]: Please remove the documentation in the section above for PRODUCT (since it duplicates this table) and add documentation to the table for RESOURCE.

Comment (melinath) [reply docs/content/docs/getting-started/generate-providers.md:38]: I'm not sure what best to do with the example of a nonsense product - maybe the more complex examples could go in a separate "examples" section after the command reference?

File: docs/layouts/shortcodes/build_table.html [hunk 14]
```diff
@@ -0,0 +1,37 @@
+<style type="text/css">
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:1]: I think this should probably be a list rather than a table based on https://developers.google.com/style/tables

**Review by melinath — CHANGES_REQUESTED** [review]

File: docs/content/docs/getting-started/generate-providers.md (hunk 13 above)
Comment (ScottSuarez) [reply docs/content/docs/getting-started/generate-providers.md]: ack, refactored the page, let me know what you think

**Comment by ScottSuarez** [comment]
I tried both table and markdown presentation for the build variable information. I still find table presentation to be more legible. I updated the table to make use of columns so we fit within the specifications lay'd out in the google stylesheet. Let me know what you think! 

Screen Shot 2022-12-29 at 4 15 50 PM
Screen Shot 2022-12-29 at 4 14 31 PM

File: docs/layouts/shortcodes/build_table.html [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+  .tg  {border-collapse:collapse;border-spacing:0;}
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:2]: please use meaningful or generic class names https://google.github.io/styleguide/htmlcssguide.html#ID_and_Class_Naming

File: docs/layouts/shortcodes/build_table.html [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+  .tg td{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
+    overflow:hidden;padding:10px 5px;word-break:normal;}
+  .tg th{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
+    font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
+  .tg .tg-c3ow{border-color:black;text-align:center;vertical-align:top}
+  .tg .tg-0pky{border-color:black;text-align:left;vertical-align:top}
+  .tg .tg-dgl5{border-color:black;background-color:#FFF;text-align:left;vertical-align:top}
+  </style>
+  <table class="tg">
+    <tr>
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:12]: This row should go inside a `<thead>` element, and the rest should go inside a `<tbody>`

**Review by melinath — CHANGES_REQUESTED** [review]
> I tried both table and markdown presentation for the build variable information. I still find table presentation to be more legible. I updated the table to make use of columns so we fit within the specifications lay'd out in the google stylesheet. Let me know what you think!

Looking through the style guide - it doesn't seem like there's clear guidance on this, actually. APIs use tables; gcloud uses man-page styles. Let's go with the table for now. 👍 

That said - please go through the table and reformat with the following guidelines:

- minimal semantic HTML - for example, use `<th>` rather than `<td><span>`, and use `<strong>` rather than span if emphasizing a single word. https://developers.google.com/style/semantic-tagging
- meaningful or generic CSS class names only if necessary - for example, put a class on the header, 
- Move the "optional" and "required" into the text on the right. (I think this will be more readable / matches API docs, but I could be convinced otherwise.)

File: docs/layouts/shortcodes/build_table.html [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+      <td class="tg-c3ow" colspan="2"><span style="font-weight:bold">magic-modules - build variables</span></td>
+    </tr>
+    <tr>
+      <td class="tg-0pky" ><span style="font-weight:bold">OUTPUT_PATH </span>(required) </td>
+      <td class="tg-0pky">The location of the provider you are unioning generation into.</td>
```
[the remark this answers is not in the document]
Comment (ScottSuarez) [reply docs/layouts/shortcodes/build_table.html]: I've decided to keep optional and required in the left hand side. It consolidates the variable and priority.. I tried both ways and did not like the feel of the other.

**Comment by ScottSuarez** [comment]
Thanks for being so thorough !! I tried to go through and rework the whole table. Admittedly the initial code was from a markdown converter tool. Definitely needed to be cleaned.

File: docs/layouts/shortcodes/build_table.html [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+        <th>OUTPUT_PATH <span class="reduce-impact">(required)</span> </th>
+        <td >The location of the provider you are unioning generation into.</td>
+      </tr>
+      <tr>
+        <th>VERSION <span class="reduce-impact">(required)</span></th>
+        <td>The version of the provider you are building into. <br/> Valid values are <span class="code">'ga'</span> and <span class="code">'beta'</span>.</td>
+      </tr>
+      <tr>
+        <th >ENGINE <span class="reduce-impact">(optional)</span></td>
+        <td >Modifies make provider to build either tpgtools or mmv1 valid values are <span class="code">'mmv1'</span> or <span class="code">`tpgtools`</span>. Note that running make provider with ENGINE set to <span class="code">`tpgtools`</span> will build any prerequisite  mmv1 files required for tpgtools.</td>
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:24]: There are extraneous code markers throughout - possibly left over from the markdown conversion? It would be great to clean up to avoid any confusion.

```suggestion
        <td >Modifies make provider to build either tpgtools or mmv1 valid values are <span class="code">mmv1</span> or <span class="code">tpgtools</span>. Note that running make provider with ENGINE set to <span class="code">tpgtools</span> will build any prerequisite  mmv1 files required for tpgtools.</td>
```

File: docs/layouts/shortcodes/build_table.html [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+      <tr>
+        <th>OUTPUT_PATH <span class="reduce-impact">(required)</span> </th>
+        <td >The location of the provider you are unioning generation into.</td>
+      </tr>
+      <tr>
+        <th>VERSION <span class="reduce-impact">(required)</span></th>
+        <td>The version of the provider you are building into. <br/> Valid values are <span class="code">'ga'</span> and <span class="code">'beta'</span>.</td>
+      </tr>
+      <tr>
+        <th >ENGINE <span class="reduce-impact">(optional)</span></td>
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:23]: There are some extraneous spaces in the opening HTML tags left over from when these had classes

```suggestion
        <th>ENGINE <span class="reduce-impact">(optional)</span></td>
```

File: docs/layouts/shortcodes/build_table.html [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+      </tr>
+    </thead>
+    <tbody>
+      <tr>
+        <th>OUTPUT_PATH <span class="reduce-impact">(required)</span> </th>
+        <td >The location of the provider you are unioning generation into.</td>
+      </tr>
+      <tr>
+        <th>VERSION <span class="reduce-impact">(required)</span></th>
+        <td>The version of the provider you are building into. <br/> Valid values are <span class="code">'ga'</span> and <span class="code">'beta'</span>.</td>
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:20]: We should use a `<code>` element in this HTML instead https://developers.google.com/style/code-in-text

Comment (melinath) [reply docs/layouts/shortcodes/build_table.html:20]:
```suggestion
        <td>The version of the provider you are building into. <br/> Valid values are <code>ga</code> and <code>beta</code>.</td>
```

File: docs/layouts/shortcodes/build_table.html [hunk 21]
```diff
@@ -0,0 +1,35 @@
+<style type="text/css">
+  .build-table  {border-collapse:collapse;border-spacing:0;}
+  .build-table th{font-weight: bold;}
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:3]: I'm a little surprised that the font-weight for `th` isn't already bold?

File: docs/layouts/shortcodes/build_table.html [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+  .build-table th span.reduce-impact{font-weight: normal;}
+  .build-table span.code{color:#905;background-color:#DDD}
```
Comment (melinath) [inline docs/layouts/shortcodes/build_table.html:5]:
```suggestion
```

If we want to change the style for `<code>` elements we should do it at the theme level rather than just for this table - it might also make sense to do in a separate PR.

**Review by melinath — CHANGES_REQUESTED** [review]
thanks for cleaning up the table - it looks a lot better! Just a few more things.

File: docs/layouts/shortcodes/build_table.html (hunk 22 above)
Comment (ScottSuarez) [reply docs/layouts/shortcodes/build_table.html]: done ! great idea... cleaning up this to rely on the default styling makes a lot more sense!

File: docs/layouts/shortcodes/build_table.html (hunk 21 above)
Comment (ScottSuarez) [reply docs/layouts/shortcodes/build_table.html]: You're correct!

File: docs/layouts/shortcodes/build_table.html (hunk 20 above)
Comment (ScottSuarez) [reply docs/layouts/shortcodes/build_table.html]: done !

**Comment by ScottSuarez** [comment]
Screen Shot 2023-01-10 at 12 51 15 PM

Great ideas migrating back to the default styling from the template. That makes way more sense then overriding with our own implicit styling system. Let me know what you think !

**Review by melinath — APPROVED** [review]
looks great, thanks for taking the time to clean up the table :-) :shipit:
